test/distributed/test_c10d.py

```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
             x = checkpoint(self.l2, x)
             return x
 
-    def _test_ddp_checkpointing(self, checkpoint_once, process_group, use_bucket_view, find_unused_parameters=False):
+    def _prepare_dummy_data(self):
+        ddp_bs = 16
+        bs = ddp_bs * self.world_size
+        input = torch.rand((bs, 20), device="cuda", requires_grad=True)
+        target = torch.randn((bs, 20), device="cuda")
+        offset = self.rank * ddp_bs
+        ddp_input = input[offset: offset + ddp_bs]
+        ddp_target = target[offset: offset + ddp_bs]
+        return input, ddp_input, target, ddp_target
+
+    def _train_model(self, model, input_var, target, loss, run_checkpoint=False):
+        model.train()
+        if run_checkpoint:
+            output = checkpoint(model, input_var)
+        else:
+            output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+
+    def _test_ddp_checkpointing(
+        self,
+        input_model,
+        process_group,
+        use_bucket_view,
+        find_unused_parameters=False,
+        static_graph=False,
+        run_checkpoint=False
+    ):
         # to reprodce the same training results
         torch.cuda.set_device(self.rank)
         torch.manual_seed(31415)
-        if checkpoint_once:
-            model = self.CheckpointOnceModule().cuda()
-        else:
-            model = self.CheckpointTwiceModule().cuda()
-        model = nn.parallel.DistributedDataParallel(model,
-                                                    bucket_cap_mb=1,
-                                                    gradient_as_bucket_view=use_bucket_view,
-                                                    device_ids=[self.rank],
-                                                    process_group=process_group,
-                                                    find_unused_parameters=find_unused_parameters)
-        input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
-        output_tensor = model(input_tensor)
-        output_tensor.sum().backward()
-        return model
+        model = copy.deepcopy(input_model).cuda()
```

Comment:
Seems like we don't need this deepcopy (can keep the next one)?

Reply:
we need this one, this is for local_model, next one is for ddp_model, we need to compare their grads